dask/dataframe/multi.py

```diff
@@ -325,6 +325,30 @@ def single_partition_join(left, right, **kwargs):
     return new_dd_object(graph, name, meta, divisions)
 
 
+def warn_dtype_mismatch(left, right, left_on, right_on):
+    """ Checks for merge column dtype mismatches and throws a warning (#4574)
+    """
+
+    if not isinstance(left_on, list):
+        left_on = [left_on]
+    if not isinstance(right_on, list):
+        right_on = [right_on]
+
+    if (all(col in left.columns for col in left_on) and
+            all(col in right.columns for col in right_on)):
+        dtype_mism = {(lo, ro): (left.dtypes[lo], right.dtypes[ro])
+                      for lo, ro in zip(left_on, right_on)
+                      if not left.dtypes[lo] is right.dtypes[ro]}
+
+        if dtype_mism:
+            col_str = '\n'.join('{}: {}'.format(cols, dtypes)
+                                for cols, dtypes in dtype_mism.items())
```

Comment:
You might consider using `dask.utils.asciitable` here to do the formatting, it's used a few other places for errors like this (e.g. https://github.com/dask/dask/pull/2522).

Reply:
Love it. I added a commit.